File: src/services/code.service.ts
import { Code } from '../types';

export class CodeService {
  /**
   * 從文本中提取驗證碼
   */
  extractCodes(text: string): Code[] {
    const codes: Code[] = [];

    // 1. 純數字驗證碼 (4-8位)
    const numericPatterns = [
      { pattern: /\b\d{6}\b/g, type: 'numeric' as const, length: 6, confidence: 0.9 },
      { pattern: /\b\d{4}\b/g, type: 'numeric' as const, length: 4, confidence: 0.8 },
      { pattern: /\b\d{8}\b/g, type: 'numeric' as const, length: 8, confidence: 0.85 },
    ];

    for (const { pattern, type, length, confidence } of numericPatterns) {
      const matches = text.match(pattern);
      if (matches) {
        matches.forEach((code) => {
          if (!this.isDuplicate(codes, code)) {
            codes.push({
              code,
              type,
              length,
              pattern: pattern.source,
              confidence,
            });
          }
        });
      }
    }

    // 2. 字母數字混合 (6-10位)
    const alphanumericPatterns = [
      {
        pattern: /\b[A-Z0-9]{6,10}\b/g,
        type: 'alphanumeric' as const,
        confidence: 0.75,
      },
    ];

    for (const { pattern, type, confidence } of alphanumericPatterns) {
      const matches = text.match(pattern);
      if (matches) {
        matches.forEach((code) => {
          if (!this.isDuplicate(codes, code)) {
            codes.push({
              code,
              type,
              length: code.length,
              pattern: pattern.source,
              confidence,
            });
          }
        });
      }
    }

    // 3. 常見驗證碼關鍵詞附近
    const contextPatterns = [
      /(?:code|Code|驗證碼|验证码|OTP|otp)[\s:：]*([A-Z0-9]{4,10})/gi,
      /(?:your|Your)\s+(?:verification|code)[\s:：]*([A-Z0-9]{4,10})/gi,
      /(?:token|Token)[\s:：]*([A-Za-z0-9_-]{10,40})/gi,
    ];

    for (const pattern of contextPatterns) {
      let match;
      while ((match = pattern.exec(text)) !== null) {
        const code = match[1];
        if (!this.isDuplicate(codes, code)) {
          const type = code.length > 15 ? 'token' : 'alphanumeric';
          codes.push({
            code,
            type: type as 'alphanumeric' | 'token',
            length: code.length,
            pattern: pattern.source,
            confidence: 0.95, // 上下文關鍵詞提高置信度
          });
        }
      }
    }

    // 4. URL參數中的驗證碼
    const urlPatterns = [
      /[?&](?:code|token|verify)=([A-Za-z0-9_-]+)/gi,
    ];

    for (const pattern of urlPatterns) {
      let match;
      while ((match = pattern.exec(text)) !== null) {
        const code = match[1];
        if (!this.isDuplicate(codes, code) && code.length >= 6) {
          codes.push({
            code,
            type: code.length > 15 ? 'token' : 'alphanumeric',
            length: code.length,
            pattern: pattern.source,
            confidence: 0.85,
          });
        }
      }
    }

    // 按置信度排序
    return codes.sort((a, b) => b.confidence - a.confidence);
  }

  /**
   * 檢查是否重複
   */
  private isDuplicate(codes: Code[], code: string): boolean {
    return codes.some((c) => c.code === code);
  }

  /**
   * 從HTML中提取驗證碼
   */
  extractFromHtml(html: string): Code[] {
    // 移除HTML標籤
    const text = html.replace(/<[^>]*>/g, ' ');
    // 解碼HTML實體
    const decoded = this.decodeHtmlEntities(text);
    return this.extractCodes(decoded);
  }

  /**
   * 解碼HTML實體
   */
  private decodeHtmlEntities(text: string): string {
    const entities: Record<string, string> = {
      '&amp;': '&',
      '&lt;': '<',
      '&gt;': '>',
      '&quot;': '"',
      '&#39;': "'",
      '&nbsp;': ' ',
    };

    return text.replace(/&[^;]+;/g, (entity) => entities[entity] || entity);
  }
}

export const codeService = new CodeService();

File: app/services/code_service.py
import re
from typing import List
from app.models import Code


class CodeService:
    """验证码提取服务"""

    def extract_codes(self, text: str) -> List[Code]:
        """从文本中提取验证码"""
        codes: List[Code] = []

        # 1. 纯数字验证码 (4-8位)
        numeric_patterns = [
            (r"\b\d{6}\b", "numeric", 6, 0.9),
            (r"\b\d{4}\b", "numeric", 4, 0.8),
            (r"\b\d{8}\b", "numeric", 8, 0.85),
        ]

        for pattern, type_, length, confidence in numeric_patterns:
            matches = re.findall(pattern, text)
            for code in matches:
                if not self._is_duplicate(codes, code):
                    codes.append(
                        Code(
                            code=code,
                            type=type_,
                            length=length,
                            pattern=pattern,
                            confidence=confidence,
                        )
                    )

        # 2. 字母數字混合 (6-10位)
        alphanumeric_pattern = r"\b[A-Z0-9]{6,10}\b"
        matches = re.findall(alphanumeric_pattern, text)
        for code in matches:
            if not self._is_duplicate(codes, code):
                codes.append(
                    Code(
                        code=code,
                        type="alphanumeric",
                        length=len(code),
                        pattern=alphanumeric_pattern,
                        confidence=0.75,
                    )
                )

        # 3. 常见验证码关键词附近
        context_patterns = [
            r"(?:code|Code|驗證碼|验证码|OTP|otp)[\s:：]*([A-Z0-9]{4,10})",
            r"(?:your|Your)\s+(?:verification|code)[\s:：]*([A-Z0-9]{4,10})",
            r"(?:token|Token)[\s:：]*([A-Za-z0-9_-]{10,40})",
        ]

        for pattern in context_patterns:
            matches = re.findall(pattern, text, re.IGNORECASE)
            for code in matches:
                if not self._is_duplicate(codes, code):
                    type_ = "token" if len(code) > 15 else "alphanumeric"
                    codes.append(
                        Code(
                            code=code,
                            type=type_,
                            length=len(code),
                            pattern=pattern,
                            confidence=0.95,  # 上下文关键词提高置信度
                        )
                    )

        # 4. URL参数中的验证码
        url_pattern = r"[?&](?:code|token|verify)=([A-Za-z0-9_-]+)"
        matches = re.findall(url_pattern, text, re.IGNORECASE)
        for code in matches:
            if not self._is_duplicate(codes, code) and len(code) >= 6:
                type_ = "token" if len(code) > 15 else "alphanumeric"
                codes.append(
                    Code(
                        code=code,
                        type=type_,
                        length=len(code),
                        pattern=url_pattern,
                        confidence=0.85,
                    )
                )

        # 按置信度排序
        return sorted(codes, key=lambda x: x.confidence, reverse=True)

    def _is_duplicate(self, codes: List[Code], code: str) -> bool:
        """检查是否重复"""
        return any(c.code == code for c in codes)

    def extract_from_html(self, html: str) -> List[Code]:
        """从HTML中提取验证码"""
        # 移除HTML标签
        text = re.sub(r"<[^>]*>", " ", html)
        # 解码HTML实体
        text = self._decode_html_entities(text)
        return self.extract_codes(text)

    def _decode_html_entities(self, text: str) -> str:
        """解码HTML实体"""
        import html

        return html.unescape(text)


# 单例
code_service = CodeService()
